Git-commit: 06ab8444392acdbffb57869d6220fb6654a8c95e
Message-Id: <06ab8444392acdbffb57869d6220fb6654a8c95e.1647486269.git.geliang.tang@suse.com>
In-Reply-To: <0d6882dd158e559b291a2d1b045a65bc2fa4fc58.1647486268.git.geliang.tang@suse.com>
References: <0d6882dd158e559b291a2d1b045a65bc2fa4fc58.1647486268.git.geliang.tang@suse.com>
From: David Howells <dhowells@redhat.com>
Date: Fri, 11 Mar 2022 13:24:29 +0000
Subject: [PATCH] watch_queue: Free the alloc bitmap when the watch_queue is
 torn down
References: bsn#19
Patch-mainline: v5.10.106

commit 7ea1a0124b6da246b5bc8c66cddaafd36acf3ecb upstream.

Free the watch_queue note allocation bitmap when the watch_queue is
destroyed.

Fixes: c73be61cede5 ("pipe: Add general notification queue support")
Reported-by: Jann Horn <jannh@google.com>
Signed-off-by: David Howells <dhowells@redhat.com>
Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Geliang Tang <geliang.tang@suse.com>
---
 kernel/watch_queue.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/kernel/watch_queue.c b/kernel/watch_queue.c
index 258fa4ca4db4..77c804d8ef7b 100644
--- a/kernel/watch_queue.c
+++ b/kernel/watch_queue.c
@@ -373,6 +373,7 @@ static void __put_watch_queue(struct kref *kref)
 
 	for (i = 0; i < wqueue->nr_pages; i++)
 		__free_page(wqueue->notes[i]);
+	bitmap_free(wqueue->notes_bitmap);
 
 	wfilter = rcu_access_pointer(wqueue->filter);
 	if (wfilter)
-- 
2.34.1

